### PR TITLE
Add `CasmContractClass`

### PR DESCRIPTION
**Pull request by giladchase**

### CODE REVIEW

**Review by elintul — CHANGES_REQUESTED** [review]
> Reviewed 9 of 13 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/445)* status: 9 of 13 files reviewed, 5 unresolved discussions (waiting on @giladchase)

___
*[`Cargo.toml` line 10 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/445#-NUa0SjAEm_T_8GZOJlv:-NUa0SjAEm_T_8GZOJlw:b-yiwgrh) ([raw file](https://github.com/starkware-libs/blockifier/blob/278c28a61cfe00abb45e7ae6ef9cae4c27d13376/Cargo.toml#L10)):*
> ```TOML
> [workspace.dependencies]
> assert_matches = { version = "1.5.0" }
> cairo-lang-starknet = "1.0.0-rc0"
> ```

Why not `{ version = ""1.0.0-rc0"" }`? Or is it not considered a version, only a release? What's the difference?
___
*[`crates/blockifier/src/execution/entry_point.rs` line 289 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/445#-NUa0qgH3hOSURim029X:-NUa0qgIE1Mwvp-JECVc:bajecej) ([raw file](https://github.com/starkware-libs/blockifier/blob/278c28a61cfe00abb45e7ae6ef9cae4c27d13376/crates/blockifier/src/execution/entry_point.rs#L289)):*
> ```XML
> ) -> EntryPointExecutionResult<CallInfo> {
>     // Ensure the class is declared (by reading it).
>     let V0(contract_class) = state.get_contract_class(&class_hash)? else {
> ```

`V0` is a bit cryptic; `ContractClass::V0`?
Planning to `match` later?
___
*[`crates/blockifier/src/execution/execution_utils.rs` line 71 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/445#-NUa18WS1GewVmGv5yB5:-NUa18WS1GewVmGv5yB6:b-q4j8qr) ([raw file](https://github.com/starkware-libs/blockifier/blob/278c28a61cfe00abb45e7ae6ef9cae4c27d13376/crates/blockifier/src/execution/execution_utils.rs#L71)):*
> ```XML
> ) -> Result<VmExecutionContext<'a>, PreExecutionError> {
>     let contract_class = state.get_contract_class(&class_hash)?;
>     match contract_class {
> ```

I think we're going to need some refactoring here, maybe before this PR?
___
*[`crates/blockifier/src/state/cached_state.rs` line 19 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/445#-NUa1IO25SaSY4_1e_tX:-NUa1IO25SaSY4_1e_tY:undefined) ([raw file](https://github.com/starkware-libs/blockifier/blob/278c28a61cfe00abb45e7ae6ef9cae4c27d13376/crates/blockifier/src/state/cached_state.rs#L19)):*
> ```XML
> 
> // TODO: add V1 mapping
> type ContractClassMapping = HashMap<ClassHash, ContractClassV0>;
> ```





_Suggestion:_
```XML
ContractClassV0Mapping
```
___
*[`crates/blockifier/src/state/cached_state.rs` line 103 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/445#-NUa1Rzv0w3dXLhEpdoX:-NUa1Rzv0w3dXLhEpdoY:bq4u3ho) ([raw file](https://github.com/starkware-libs/blockifier/blob/278c28a61cfe00abb45e7ae6ef9cae4c27d13376/crates/blockifier/src/state/cached_state.rs#L103)):*
> ```XML
>         if !self.class_hash_to_class.contains_key(class_hash) {
>             match self.state.get_contract_class(class_hash)? {
>                 ContractClass::V1(_) => todo!(),
> ```

@noaov1, for visibility.



_Code quote:_
```XML
todo!()
```

**Review by elintul — CHANGES_REQUESTED** [review]
> Reviewed 4 of 13 files at r1.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/445)* status: all files reviewed, 7 unresolved discussions (waiting on @giladchase)

___
*[`crates/blockifier/src/state/cached_state_test.rs` line 143 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/445#-NUa2Z2s8HKeUBEbB969:-NUa2Z2s8HKeUBEbB96A:b-g41kmc) ([raw file](https://github.com/starkware-libs/blockifier/blob/278c28a61cfe00abb45e7ae6ef9cae4c27d13376/crates/blockifier/src/state/cached_state_test.rs#L143)):*
> ```XML
>     let ContractClass::V0(contract_class) = state.get_contract_class(&existing_class_hash).unwrap() else {
>         panic!("Test class hash should match V0 class.")
>     };
> ```

How do I read this line?



_Code quote:_
```XML
    let ContractClass::V0(contract_class) = state.get_contract_class(&existing_class_hash).unwrap() else {
        panic!("Test class hash should match V0 class.")
    };
```
___
*[`crates/blockifier/src/transaction/transactions.rs` line 93 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/445#-NUa273g5GOHVF0erSQY:-NUa273g5GOHVF0erSQZ:b-64y3u0) ([raw file](https://github.com/starkware-libs/blockifier/blob/278c28a61cfe00abb45e7ae6ef9cae4c27d13376/crates/blockifier/src/transaction/transactions.rs#L93)):*
> ```XML
>                 match contract_class.expect("Declare transaction must have a contract_class") {
>                     ContractClass::V0(contract_class) => {
>                         state.set_contract_class(&class_hash, contract_class)?
> ```

Planning to have a single setter?



_Code quote:_
```XML
set_contract_class
```

**Review by elintul — APPROVED** [review]
:lgtm:

> Reviewed 10 of 10 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/445#-:-NUwtsQRCx9PPoGFJvOQ:bnfp4nl)* status: all files reviewed, 3 unresolved discussions (waiting on @giladchase)

___
*[`crates/blockifier/src/execution/entry_point.rs` line 289 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/445#-NUa0qgH3hOSURim029X:-NUwsX715lo5VrGE0_Hj:b-vbt41) ([raw file](https://github.com/starkware-libs/blockifier/blob/278c28a61cfe00abb45e7ae6ef9cae4c27d13376/crates/blockifier/src/execution/entry_point.rs#L289)):*
<details><summary><i>Previously, giladchase wrote…</i></summary><blockquote>

Fixed.
What do you mean about the `match`?
</blockquote></details>

Instead of the `todo!()`.
___
*[`crates/blockifier/src/execution/execution_utils.rs` line 144 at r2](https://reviewable.io/reviews/starkware-libs/blockifier/445#-NUaY8Md0SR8hsbvslyt:-NUwtpve4lqLpm35zruZ:b1rbbom) ([raw file](https://github.com/starkware-libs/blockifier/blob/d27ac1380e27688bd2d73249a9048c17da52eafa/crates/blockifier/src/execution/execution_utils.rs#L144)):*
<details><summary><i>Previously, giladchase wrote…</i></summary><blockquote>

Hoisted up one more level (is that what you meant by "callsite"? cause we already hoisted it up one level once from inside `initialize_execution_context`).

The caller where I moved it to (inside entry_point.rs) is already in multiple places (so I'll have to put this code in many places if we want this higher up in the call stack).
</blockquote></details>

Only two places, right?
___
*[`crates/blockifier/src/state/cached_state_test.rs` line 143 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/445#-NUa2Z2s8HKeUBEbB969:-NUwtFir9MsKIj2Kybv8:bbx3omy) ([raw file](https://github.com/starkware-libs/blockifier/blob/278c28a61cfe00abb45e7ae6ef9cae4c27d13376/crates/blockifier/src/state/cached_state_test.rs#L143)):*
<details><summary><i>Previously, giladchase wrote…</i></summary><blockquote>

Let-else saves you the indentation level when extracting the value from option/result in cases where only one variant is relevant, and the rest are usually short-circuited out via panic/return Err.

So basically if you have code like the following, and you don't want to pay for `foo` with indentation:
```rust
if let MyEnum::GoodVariant(foo) = get_my_enum() {
     // foo in scope, but with indentation
     do_something(foo);
     do_something_else(foo) 
} else {
   <raise error> // all other variants are bad
}
```
you can do
```rust

let MyEnum::GoodVariant(foo) = get_my_enum() else {
   <raise error> // all other variants are bad
}

// Foo is in scope, without indentation!
do_something(foo);
do_something_else(foo) 

```
is the same as:


In the former, the happy flow is indented, whereas in the latterit is not.
</blockquote></details>

מהמם

**Review by giladchase — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/445)* status: all files reviewed, 3 unresolved discussions (waiting on @elintul)

___
*[`crates/blockifier/src/execution/entry_point.rs` line 289 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/445#-NUa0qgH3hOSURim029X:-NUzl7J2BJqFYNLhxVfT:b-xjtk98) ([raw file](https://github.com/starkware-libs/blockifier/blob/278c28a61cfe00abb45e7ae6ef9cae4c27d13376/crates/blockifier/src/execution/entry_point.rs#L289)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

Instead of the `todo!()`.
</blockquote></details>

Oh, ya.
___
*[`crates/blockifier/src/execution/execution_utils.rs` line 144 at r2](https://reviewable.io/reviews/starkware-libs/blockifier/445#-NUaY8Md0SR8hsbvslyt:-NUzlAjQA-MaucOD-Ed5:b9tdd9f) ([raw file](https://github.com/starkware-libs/blockifier/blob/d27ac1380e27688bd2d73249a9048c17da52eafa/crates/blockifier/src/execution/execution_utils.rs#L144)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

Only two places, right?
</blockquote></details>

7, not including test locations (it's all places that call CallEntryPoint.execute):
https://github.com/starkware-libs/blockifier/blob/b1f04ef9c0e262e956b3aee16559b69e81a1d3d1/crates/blockifier/src/execution/deprecated_syscalls/hint_processor.rs#L398
https://github.com/starkware-libs/blockifier/blob/ddf6c113d5ad58a4145e62a11bef3d96353cd924/crates/blockifier/src/execution/entry_point.rs#L319
https://github.com/starkware-libs/blockifier/blob/b1f04ef9c0e262e956b3aee16559b69e81a1d3d1/crates/blockifier/src/execution/syscalls/hint_processor.rs#L372
https://github.com/starkware-libs/blockifier/blob/d2660e200d1b8879e89cde873007bd8605200bfd/crates/blockifier/src/transaction/account_transaction.rs#L176
https://github.com/starkware-libs/blockifier/blob/d2660e200d1b8879e89cde873007bd8605200bfd/crates/blockifier/src/transaction/account_transaction.rs#L250
https://github.com/starkware-libs/blockifier/blob/35a6ee6a51482bdb4b4440d5fb72f937f086c082/crates/blockifier/src/transaction/transactions.rs#L174
https://github.com/starkware-libs/blockifier/blob/35a6ee6a51482bdb4b4440d5fb72f937f086c082/crates/blockifier/src/transaction/transactions.rs#L208

That being said, maybe I'm still misunderstanding what you mean by `callsite`, how high up do we want to go?
___
*[`crates/blockifier/src/state/cached_state_test.rs` line 143 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/445#-NUa2Z2s8HKeUBEbB969:-NUzm8-m5ud5h6irws4n:b2x9v3l) ([raw file](https://github.com/starkware-libs/blockifier/blob/278c28a61cfe00abb45e7ae6ef9cae4c27d13376/crates/blockifier/src/state/cached_state_test.rs#L143)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

מהמם
</blockquote></details>

:blush:
